node/node.go

```diff
@@ -406,42 +420,117 @@ func updateNodeLease(ctx context.Context, leases v1beta1.LeaseInterface, lease *
+	oldVKStatus, ok1 := apiServerNode.Annotations[virtualKubeletLastNodeAppliedNodeStatus]
+	oldVKObjectMeta, ok2 := apiServerNode.Annotations[virtualKubeletLastNodeAppliedObjectMeta]
+
+	oldNode := corev1.Node{}
+	// Check if there were no labels, which means someone else probably created the node, or this is an upgrade. Either way, we will consider
+	// ourselves as never having written the node object before, so oldNode will be left empty. We will overwrite values if
+	// our new node conditions / status / objectmeta have them
+	if ok1 && ok2 {
+		err := json.Unmarshal([]byte(oldVKObjectMeta), &oldNode.ObjectMeta)
+		if err != nil {
+			return nil, pkgerrors.Wrapf(err, "Cannot unmarshal old node object metadata (key: %q): %q", virtualKubeletLastNodeAppliedObjectMeta, oldVKObjectMeta)
+		}
+		err = json.Unmarshal([]byte(oldVKStatus), &oldNode.Status)
+		if err != nil {
+			return nil, pkgerrors.Wrapf(err, "Cannot unmarshal old node status (key: %q): %q", virtualKubeletLastNodeAppliedNodeStatus, oldVKStatus)
+		}
 	}
 
-	updatedNode, err := nodes.Patch(string(nodeName), types.StrategicMergePatchType, patchBytes, "status")
+	// newNode is the representation of the node the provider "wants"
+	newNode := corev1.Node{}
+	newNode.ObjectMeta = simplestObjectMetadata(&apiServerNode.ObjectMeta, &nodeFromProvider.ObjectMeta)
+	nodeFromProvider.Status.DeepCopyInto(&newNode.Status)
+
+	// virtualKubeletLastNodeAppliedObjectMeta must always be set before virtualKubeletLastNodeAppliedNodeStatus,
+	// otherwise we capture virtualKubeletLastNodeAppliedNodeStatus in virtualKubeletLastNodeAppliedObjectMeta,
+	// which is wrong
+	virtualKubeletLastNodeAppliedObjectMetaBytes, err := json.Marshal(newNode.ObjectMeta)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to patch status %q for node %q: %v", patchBytes, nodeName, err)
+		return nil, pkgerrors.Wrap(err, "Cannot marshal object meta from provider")
 	}
-	return updatedNode, patchBytes, nil
-}
+	newNode.Annotations[virtualKubeletLastNodeAppliedObjectMeta] = string(virtualKubeletLastNodeAppliedObjectMetaBytes)
 
-func preparePatchBytesforNodeStatus(nodeName types.NodeName, oldNode *corev1.Node, newNode *corev1.Node) ([]byte, error) {
-	oldData, err := json.Marshal(oldNode)
+	virtualKubeletLastNodeAppliedNodeStatusBytes, err := json.Marshal(newNode.Status)
 	if err != nil {
-		return nil, fmt.Errorf("failed to Marshal oldData for node %q: %v", nodeName, err)
+		return nil, pkgerrors.Wrap(err, "Cannot marshal node status from provider")
 	}
-
-	// Reset spec to make sure only patch for Status or ObjectMeta is generated.
-	// Note that we don't reset ObjectMeta here, because:
-	// 1. This aligns with Nodes().UpdateStatus().
-	// 2. Some component does use this to update node annotations.
-	newNode.Spec = oldNode.Spec
-	newData, err := json.Marshal(newNode)
+	newNode.Annotations[virtualKubeletLastNodeAppliedNodeStatus] = string(virtualKubeletLastNodeAppliedNodeStatusBytes)
```

Comment:
Should these be base64 encoded?

Reply:
Nope. JSON is fine to put here. In fact, this is exactly how kubectl does it. Base64 is just adding overhead and a readability headache.